File: experiment-13/runs/language=python_model=opus-4.8-fast_prompt=ATDD/rep1/brazilian_soccer_mcp/loader.py
"""
Context
=======
CSV loaders for the Brazilian Soccer MCP server.

Each of the six provided Kaggle files has a different schema; this module parses
all of them into the unified ``Match`` / ``Player`` domain records. Loading is
done with the standard-library ``csv`` module (no heavy dependencies) and is
tolerant of the data-quality issues called out in the spec: multiple date
formats, goals stored as ints / floats / strings, UTF-8 with a leading BOM, and
team-name variations (handled via ``normalize``).

Files are detected by name, so a data directory may contain any subset of the
six files (tests seed only what they need; production has all six).
"""

from __future__ import annotations

import csv
from dataclasses import replace
from pathlib import Path
from typing import Iterable, Optional

from .models import Match, Player
from .normalize import key, normalize_team

# Canonical competition labels. Several files cover the same competition under
# different names (e.g. "Serie A" == "Brasileirão"); canonicalising lets the
# loader de-duplicate matches that appear in more than one source file.
BRASILEIRAO = "Brasileirão"
# Named without the "Brasileirão" prefix so a substring query for the top flight
# ("Brasileirão") does not accidentally include the lower divisions.
BRASILEIRAO_B = "Série B"
BRASILEIRAO_C = "Série C"
COPA_DO_BRASIL = "Copa do Brasil"
LIBERTADORES = "Copa Libertadores"

# BR-Football "tournament" values -> canonical competition.
_TOURNAMENT_ALIASES = {
    "serie a": BRASILEIRAO,
    "serie b": BRASILEIRAO_B,
    "serie c": BRASILEIRAO_C,
    "copa do brasil": COPA_DO_BRASIL,
    "copa libertadores": LIBERTADORES,
    "libertadores": LIBERTADORES,
}


def canonical_competition(raw: str) -> str:
    """Map a raw competition / tournament label onto a canonical name."""
    if not raw:
        return "Unknown"
    return _TOURNAMENT_ALIASES.get(key(raw), str(raw).strip())


def _to_int(value) -> Optional[int]:
    if value is None:
        return None
    text = str(value).strip()
    if text == "" or text.lower() in {"nan", "none", "null"}:
        return None
    try:
        return int(float(text))
    except ValueError:
        return None


def _parse_date(value) -> tuple[Optional[str], Optional[int]]:
    """Return ``(iso_date, year)`` for a raw date string in any known format."""
    if value is None:
        return None, None
    text = str(value).strip()
    if not text:
        return None, None
    # Drop a trailing time component: "2012-05-19 18:30:00".
    head = text.split(" ")[0].split("T")[0]
    # Brazilian format DD/MM/YYYY.
    if "/" in head:
        parts = head.split("/")
        if len(parts) == 3:
            d, m, y = parts
            if len(y) == 4:
                return f"{y}-{int(m):02d}-{int(d):02d}", int(y)
    # ISO format YYYY-MM-DD.
    if "-" in head:
        parts = head.split("-")
        if len(parts) == 3 and len(parts[0]) == 4:
            y, m, d = parts
            try:
                return f"{int(y):04d}-{int(m):02d}-{int(d):02d}", int(y)
            except ValueError:
                return None, None
    return None, None


def _rows(path: Path) -> Iterable[dict]:
    # utf-8-sig transparently strips a leading BOM (present in fifa_data.csv).
    with path.open("r", newline="", encoding="utf-8-sig") as fh:
        yield from csv.DictReader(fh)


def _load_brasileirao(path: Path) -> list[Match]:
    matches = []
    for row in _rows(path):
        iso, year = _parse_date(row.get("datetime"))
        matches.append(Match(
            competition=BRASILEIRAO,
            season=_to_int(row.get("season")) or year,
            date=iso,
            home_team=normalize_team(row.get("home_team")),
            away_team=normalize_team(row.get("away_team")),
            home_goal=_to_int(row.get("home_goal")),
            away_goal=_to_int(row.get("away_goal")),
            round=(str(row.get("round")).strip() or None) if row.get("round") else None,
        ))
    return matches


def _load_copa(path: Path) -> list[Match]:
    matches = []
    for row in _rows(path):
        iso, year = _parse_date(row.get("datetime"))
        matches.append(Match(
            competition=COPA_DO_BRASIL,
            season=_to_int(row.get("season")) or year,
            date=iso,
            home_team=normalize_team(row.get("home_team")),
            away_team=normalize_team(row.get("away_team")),
            home_goal=_to_int(row.get("home_goal")),
            away_goal=_to_int(row.get("away_goal")),
            round=(str(row.get("round")).strip() or None) if row.get("round") else None,
        ))
    return matches


def _load_libertadores(path: Path) -> list[Match]:
    matches = []
    for row in _rows(path):
        iso, year = _parse_date(row.get("datetime"))
        matches.append(Match(
            competition=LIBERTADORES,
            season=_to_int(row.get("season")) or year,
            date=iso,
            home_team=normalize_team(row.get("home_team")),
            away_team=normalize_team(row.get("away_team")),
            home_goal=_to_int(row.get("home_goal")),
            away_goal=_to_int(row.get("away_goal")),
            stage=(str(row.get("stage")).strip() or None) if row.get("stage") else None,
        ))
    return matches


def _load_br_football(path: Path) -> list[Match]:
    matches = []
    for row in _rows(path):
        iso, year = _parse_date(row.get("date"))
        matches.append(Match(
            competition=canonical_competition(row.get("tournament")),
            season=year,
            date=iso,
            home_team=normalize_team(row.get("home")),
            away_team=normalize_team(row.get("away")),
            home_goal=_to_int(row.get("home_goal")),
            away_goal=_to_int(row.get("away_goal")),
        ))
    return matches


def _load_historical(path: Path) -> list[Match]:
    matches = []
    for row in _rows(path):
        iso, year = _parse_date(row.get("Data"))
        matches.append(Match(
            competition=BRASILEIRAO,
            season=_to_int(row.get("Ano")) or year,
            date=iso,
            home_team=normalize_team(row.get("Equipe_mandante")),
            away_team=normalize_team(row.get("Equipe_visitante")),
            home_goal=_to_int(row.get("Gols_mandante")),
            away_goal=_to_int(row.get("Gols_visitante")),
            round=(str(row.get("Rodada")).strip() or None) if row.get("Rodada") else None,
        ))
    return matches


def _load_players(path: Path) -> list[Player]:
    players = []
    for row in _rows(path):
        name = (row.get("Name") or "").strip()
        if not name:
            continue
        players.append(Player(
            name=name,
            age=_to_int(row.get("Age")),
            nationality=(row.get("Nationality") or "").strip(),
            overall=_to_int(row.get("Overall")),
            potential=_to_int(row.get("Potential")),
            club=normalize_team(row.get("Club")) if row.get("Club") else "",
            position=(row.get("Position") or "").strip(),
        ))
    return players


_MATCH_FILES = {
    "Brasileirao_Matches.csv": _load_brasileirao,
    "Brazilian_Cup_Matches.csv": _load_copa,
    "Libertadores_Matches.csv": _load_libertadores,
    "BR-Football-Dataset.csv": _load_br_football,
    "novo_campeonato_brasileiro.csv": _load_historical,
}


def _select_primary_source(matches: list[Match]) -> list[Match]:
    """Resolve overlapping source files so each competition-season is consistent.

    Several files cover the same competition and years (e.g. Brasileirão Série A
    appears in three files), spelling teams differently. Unioning them would
    invent phantom fixtures (same game, two spellings) and inflate standings. So
    for each (competition, season) only the single source file contributing the
    most matches is kept — yielding one clean, internally-consistent record set
    per season while still drawing on whichever file covers each season best.
    """
    from collections import Counter, defaultdict

    groups: dict[tuple, list[Match]] = defaultdict(list)
    for m in matches:
        groups[(key(m.competition), m.season)].append(m)

    result: list[Match] = []
    for group in groups.values():
        counts = Counter(m.source for m in group)
        best = counts.most_common(1)[0][0]
        # Within the chosen source, keep one record per ordered pairing.
        seen = set()
        for m in group:
            if m.source != best:
                continue
            pairing = (key(m.home_team), key(m.away_team))
            if pairing in seen:
                continue
            seen.add(pairing)
            result.append(m)
    return result


def load_matches(data_dir: Path) -> list[Match]:
    matches: list[Match] = []
    for filename, parser in _MATCH_FILES.items():
        path = data_dir / filename
        if path.exists():
            matches.extend(
                replace(m, source=filename) for m in parser(path)
            )
    return _select_primary_source(matches)


def load_players(data_dir: Path) -> list[Player]:
    path = data_dir / "fifa_data.csv"
    return _load_players(path) if path.exists() else []

File: experiment-26-brazil-35b-60m/brazil/runs/agent=hermes-local_language=go_prompt=none/rep2/loader.go
// The CSV loader reads and normalizes team names across all 6 Kaggle datasets,
// parsing dates and numeric fields into Go types for storage in the MCP store.
package main

import (
	"encoding/csv"
	"fmt"
	"io"
	"os"
	"strconv"
	"strings"
)

// LoadBrasileirao loads Brasileirao Serie A matches (home_team_state suffix).
func LoadBrasileirao(path string) ([]Match, error) {
	f, err := os.Open(path)
	if err != nil {
		return nil, fmt.Errorf("open brasileirao: %w", err)
	}
	defer f.Close()

	var matches []Match
	r := csv.NewReader(f)
	r.LazyQuotes = true
	_, err = r.Read() // skip header
	if err != nil {
		return nil, fmt.Errorf("read header: %w", err)
	}

	for {
		row, err := r.Read()
		if err == io.EOF {
			break
		}
		if err != nil {
			return nil, fmt.Errorf("read row: %w", err)
		}
		if len(row) < 9 {
			continue
		}

		season := parseInt(strings.TrimSpace(row[7]))
		round := parseInt(strings.TrimSpace(row[8]))

		matches = append(matches, Match{
			DateTime:    strings.TrimSpace(row[0]),
			HomeTeam:    normalizeTeamName(strings.TrimSpace(row[1])),
			AwayTeam:    normalizeTeamName(strings.TrimSpace(row[4])),
			HomeScore:   parseInt(row[5]),
			AwayScore:   parseInt(row[6]),
			Season:      season,
			Round:       strconv.Itoa(round),
			Competition: "Brasileirao Serie A",
			Source:      "Brasileirao_Matches.csv",
		})
	}
	return matches, nil
}

// LoadCopaBrasil loads Copa do Brasil matches.
func LoadCopaBrasil(path string) ([]Match, error) {
	f, err := os.Open(path)
	if err != nil {
		return nil, fmt.Errorf("open copa brasil: %w", err)
	}
	defer f.Close()

	var matches []Match
	r := csv.NewReader(f)
	r.LazyQuotes = true
	_, err = r.Read()
	if err != nil {
		return nil, fmt.Errorf("read header: %w", err)
	}

	for {
		row, err := r.Read()
		if err == io.EOF {
			break
		}
		if err != nil {
			return nil, fmt.Errorf("read row: %w", err)
		}
		if len(row) < 7 {
			continue
		}

		season := parseInt(strings.TrimSpace(row[6]))
		round := strings.TrimSpace(row[0])

		matches = append(matches, Match{
			DateTime:    strings.TrimSpace(row[1]),
			HomeTeam:    normalizeTeamName(strings.TrimSpace(row[2])),
			AwayTeam:    normalizeTeamName(strings.TrimSpace(row[3])),
			HomeScore:   parseInt(row[4]),
			AwayScore:   parseInt(row[5]),
			Season:      season,
			Round:       round,
			Competition: "Copa do Brasil",
			Source:      "Brazilian_Cup_Matches.csv",
		})
	}
	return matches, nil
}

// LoadLibertadores loads Copa Libertadores matches.
func LoadLibertadores(path string) ([]Match, error) {
	f, err := os.Open(path)
	if err != nil {
		return nil, fmt.Errorf("open libertadores: %w", err)
	}
	defer f.Close()

	var matches []Match
	r := csv.NewReader(f)
	r.LazyQuotes = true
	_, err = r.Read()
	if err != nil {
		return nil, fmt.Errorf("read header: %w", err)
	}

	for {
		row, err := r.Read()
		if err == io.EOF {
			break
		}
		if err != nil {
			return nil, fmt.Errorf("read row: %w", err)
		}
		if len(row) < 7 {
			continue
		}

		season := parseInt(strings.TrimSpace(row[5]))

		matches = append(matches, Match{
			DateTime:    strings.TrimSpace(row[0]),
			HomeTeam:    normalizeTeamName(strings.TrimSpace(row[1])),
			AwayTeam:    normalizeTeamName(strings.TrimSpace(row[2])),
			HomeScore:   parseInt(row[3]),
			AwayScore:   parseInt(row[4]),
			Season:      season,
			Stage:       strings.TrimSpace(row[6]),
			Competition: "Copa Libertadores",
			Source:      "Libertadores_Matches.csv",
		})
	}
	return matches, nil
}

// LoadBRFootball loads the extended match statistics dataset.
func LoadBRFootball(path string) ([]Match, error) {
	f, err := os.Open(path)
	if err != nil {
		return nil, fmt.Errorf("open br-football: %w", err)
	}
	defer f.Close()

	var matches []Match
	r := csv.NewReader(f)
	r.LazyQuotes = true
	_, err = r.Read()
	if err != nil {
		return nil, fmt.Errorf("read header: %w", err)
	}

	for {
		row, err := r.Read()
		if err == io.EOF {
			break
		}
		if err != nil {
			return nil, fmt.Errorf("read row: %w", err)
		}
		if len(row) < 18 {
			continue
		}

		// Extract date from the "date" column (index 12)
		dateStr := strings.TrimSpace(row[12])
		// Extract season from date if possible
		season := 0
		if len(dateStr) >= 4 {
			parts := strings.Split(dateStr, "-")
			if len(parts) >= 1 {
				season = parseInt(parts[0])
			}
		}

		tournament := strings.TrimSpace(row[0])
		competition := tournament
		if competition == "" {
			competition = "Brazilian Football"
		}

		matches = append(matches, Match{
			DateTime:    dateStr,
			HomeTeam:    normalizeTeamName(strings.TrimSpace(row[1])),
			AwayTeam:    normalizeTeamName(strings.TrimSpace(row[4])),
			HomeScore:   parseInt(row[2]),
			AwayScore:   parseInt(row[3]),
			Season:      season,
			Tournament:  tournament,
			Competition: competition,
			Source:      "BR-Football-Dataset.csv",
		})
	}
	return matches, nil
}

// LoadNovoCampeonato loads historical Brasileirao 2003-2019.
func LoadNovoCampeonato(path string) ([]Match, error) {
	f, err := os.Open(path)
	if err != nil {
		return nil, fmt.Errorf("open novo campeonato: %w", err)
	}
	defer f.Close()

	var matches []Match
	r := csv.NewReader(f)
	r.LazyQuotes = true
	_, err = r.Read()
	if err != nil {
		return nil, fmt.Errorf("read header: %w", err)
	}

	for {
		row, err := r.Read()
		if err == io.EOF {
			break
		}
		if err != nil {
			return nil, fmt.Errorf("read row: %w", err)
		}
		if len(row) < 13 {
			continue
		}

		// Date format: DD/MM/YYYY
		dateStr := strings.TrimSpace(row[1])
		year := 0
		if len(dateStr) >= 7 {
			parts := strings.Split(dateStr, "/")
			if len(parts) >= 3 {
				year = parseInt(parts[2])
			}
		}

		round := parseInt(strings.TrimSpace(row[3]))

		matches = append(matches, Match{
			DateTime:    dateStr,
			HomeTeam:    normalizeTeamName(strings.TrimSpace(row[4])),
			AwayTeam:    normalizeTeamName(strings.TrimSpace(row[5])),
			HomeScore:   parseInt(row[6]),
			AwayScore:   parseInt(row[7]),
			Season:      year,
			Round:       strconv.Itoa(round),
			Competition: "Campeonato Brasileiro",
			Source:      "novo_campeonato_brasileiro.csv",
		})
	}
	return matches, nil
}

// LoadFIFAPlayers loads FIFA player database.
func LoadFIFAPlayers(path string) ([]Player, error) {
	f, err := os.Open(path)
	if err != nil {
		return nil, fmt.Errorf("open fifa: %w", err)
	}
	defer f.Close()

	var players []Player
	r := csv.NewReader(f)
	r.LazyQuotes = true
	header, err := r.Read()
	if err != nil {
		return nil, fmt.Errorf("read header: %w", err)
	}

	// Find column indices (header[0] may be empty due to BOM)
	colIdx := make(map[string]int)
	for i, name := range header {
		colIdx[strings.TrimSpace(name)] = i
	}

	for {
		row, err := r.Read()
		if err == io.EOF {
			break
		}
		if err != nil {
			return nil, fmt.Errorf("read row: %w", err)
		}
		if len(row) < 10 {
			continue
		}

		id := parseInt(getCol(row, colIdx, "ID"))
		name := strings.TrimSpace(getCol(row, colIdx, "Name"))
		age := parseInt(getCol(row, colIdx, "Age"))
		nationality := strings.TrimSpace(getCol(row, colIdx, "Nationality"))
		overall := parseInt(getCol(row, colIdx, "Overall"))
		potential := parseInt(getCol(row, colIdx, "Potential"))
		club := strings.TrimSpace(getCol(row, colIdx, "Club"))
		position := strings.TrimSpace(getCol(row, colIdx, "Position"))

		players = append(players, Player{
			ID:          id,
			Name:        name,
			Age:         age,
			Nationality: nationality,
			Overall:     overall,
			Potential:   potential,
			Club:        club,
			Position:    position,
		})
	}
	return players, nil
}

// Helper functions
func normalizeTeamName(name string) string {
	// Remove state suffixes like -SP, -RJ, -MG, -PR, -RS, -BA, -CE, -GO, -PE, -AL, -MT, -PA, -PB, -ES, -DF
	for _, suffix := range []string{"-SP", "-RJ", "-MG", "-PR", "-RS", "-BA", "-CE", "-GO", "-PE", "-AL", "-MT", "-PA", "-PB", "-ES", "-DF"} {
		name = strings.TrimSuffix(name, suffix)
	}
	// Trim whitespace and clean up
	name = strings.TrimSpace(name)
	return name
}

func parseInt(s string) int {
	s = strings.TrimSpace(s)
	s = strings.ReplaceAll(s, ".", "")
	val, err := strconv.Atoi(s)
	if err != nil {
		return 0
	}
	return val
}

func getCol(row []string, idx map[string]int, name string) string {
	if i, ok := idx[name]; ok && i < len(row) {
		return row[i]
	}
	if i, ok := idx[""+name]; ok && i < len(row) {
		return row[i]
	}
	return ""
}
